### ISSUE TITLE
When using pnpm: Cannot find package 'shiki' imported

### ISSUE THREAD

**Issue by Yue-plus** [withastro/astro#6059]
### What version of `astro` are you using?

2.0.4

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

pnpm

### What operating system are you using?

Windows 11 22H2

### Describe the Bug

```text
PS D:\code\astro-endfield> pnpm build

> astro-endfield@0.0.0 build D:\code\astro-endfield
> astro build

20:00:10 [build] output target: static
20:00:10 [build] Collecting build info...
20:00:10 [build] Completed in 40ms.
20:00:10 [build] Building static entrypoints...
The language "coffeescript" doesn't exist, falling back to plaintext.
The language "c++" doesn't exist, falling back to plaintext.
The language "cs" doesn't exist, falling back to plaintext.
20:00:13 [build] Completed in 3.30s.

 building client 
Completed in 332ms.


 generating static routes 
 error   Cannot find package 'shiki' imported from D:\code\astro-endfield\dist\entry.mjs
  Did you mean to import shiki@0.11.1/node_modules/shiki/dist/index.js?
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'shiki' imported from D:\code\astro-endfield\dist\entry.mjs
Did you mean to import shiki@0.11.1/node_modules/shiki/dist/index.js?
    at new NodeError (node:internal/errors:399:5)
    at packageResolve (node:internal/modules/esm/resolve:794:9)
    at moduleResolve (node:internal/modules/esm/resolve:843:20)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/loader:164:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:419:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36)
 ELIFECYCLE  Command failed with exit code 1.
PS D:\code\astro-endfield> 
```

### Link to Minimal Reproducible Example

https://github.com/Yue-plus/astro-endfield

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Yue-plus** [comment]
There is no such problem when using Yarn package manager.

**Comment by Yue-plus** [comment]
There is no such problem when using npm package manager.

**Comment by matthewp** [comment]
You need to use something like hoist-pattern so that shiki is hoisted: https://pnpm.io/npmrc#hoist-pattern

**Comment by petermekhaeil** [comment]
I didn't have this problem on my recent Astro app. Perhaps its because I had to also install shiki for my use-case.

https://github.com/petermekhaeil/petermekhaeil.com/blob/master/package.json

**Comment by matthewp** [comment]
Yes, the reason is because shiki cannot be bundled, so it must be marked as `external` which causing the hoisting requirement. If you had shiki yourself you wouldn't run into the issue.

**Comment by rstacruz** [comment]
Hey there - just leaving a comment for anyone finding this through a search.

I encountered this same error when using Yarn v1.

> Cannot find package 'shiki' imported from...

I migrated the project from Yarn v1 to npm@9 and found that there were some conflicting packages in my project (eg, I updated from Astro 2 to Astro 3, and there were some packages that had astro 2 as a peer dependency).

I fixed those and the problem went away :sparkles:
